Extract SendNode#access_modifier? predicate

backus opened the pull request:
Expands the `#macro?` predicate to the following cases:

  1. Class constructors:

      ```ruby
      Module.new do
        validate :name
      end
      ```

  2. Singleton classes:

      ```ruby
      class << self
        def foo
        end

        private

        def bar
        end
      end
      ```

  3. Any block:

      ```ruby
      concern :Auth do
        def foo
        end

        private

        def bar
        end
      end
      ```

  4. A node within an explicit `begin ... end` but otherwise within a
     valid macro scope:

     ```ruby
     class Foo
       begin
         validate :foo
       end
     end
     ```

  5. A node called in a root scope:

     ```ruby
     task :foo
     ```

Replaces the `AccessModifierNode` mixin with a predicate method for
method dispatch nodes.

Only `public`, `protected`, `private` and `module_function` are
respected right now. Methods like `private_constant` cannot be
called without arguments right now and were not the concern of any
of `AccessModifierNode` before. `SendNode` should support these
also eventually but it doesn't seem appropriate for this PR.

I've updated the rubocop_todo since I made the largest class one line longer with my change. A few of the offense counts have changed since it was last updated so they're in the diff as well

---

I've edited this PR description since I first opened. If you want to read the original intermediary description for context:

<details>
I started extracting `ClassNode`, `ModuleNode`, and `SingletonClassNode` and found that I really wanted an `#access_modifier?` predicate.

I've got the predicate extracted and almost all of the code is passing. There is a small amount of code though that seems to almost be an accident that enables things like

```ruby
concern do
  def foo
  end

  private

  def bar
  end
end
```

to be handled properly. I could have just encoded this in as well but I wanted a second opinion before I just baked it into the logic of `#macro?`. WDYT @drenmi?
</details>

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

backus commented:
BTW code isn't up to snuff otherwise I just wanted to queue up some input on the whole "should blocks be treated as valid for class macros?" question

Drenmi commented on lib/rubocop/ast/node/mixin/method_dispatch_node.rb:
```diff
@@ -8,6 +8,8 @@ module MethodDispatchNode
       extend NodePattern::Macros
       include MethodIdentifierPredicates
 
+      ACCESS_MODIFIERS = %i[public protected private module_function].freeze
```
There's also `#private_class_method`, `#public_class_method`, `#private_constant`, and `#public_constant`. 🙂

backus replied:
True. I was thinking it should only be for the bare method calls since it is used for things like rails indentation behavior. I'll see if it works to add these in

backus replied:
I don't think they belong in this PR. They weren't covered by any of the cops that I updated and adding support for them would require adding other public API to the send nodes.

Drenmi replied:
@backus: There's a bit of a backlog on the new 2.1 style modifiers. Most of our indentation cops don't handle them at this point. 😞 

That said, I can add the differentiating predicates in when I work on those. No need to get it in with this PR, since it's not related. 🙂

Drenmi commented on lib/rubocop/ast/node/mixin/method_dispatch_node.rb:
```diff
@@ -130,8 +136,8 @@ def def_modifier?
       private
 
       def_node_matcher :macro_scope?, <<-PATTERN
```
🎉

Drenmi commented on spec/rubocop/ast/send_node_spec.rb:
```diff
@@ -94,6 +94,35 @@
+        end
+      RUBY
+
+      it 'classifies specific macros like `private` as access modifiers' do
```
I really prefer the (very repetitive) style used in the rest of the spec, with one `it` block per assertion, encapsulating the source. 🙂

backus replied:
I think the RSpec core guys view short non-descriptive specs as no longer a good thing. Hence them removing the bare ` it { should }` and recommending you not use `is_expected.to`.

That is just a note though. I'll almost always support in-file consistency over other high level things like that so will change :smile:

Drenmi commented:
Looks like a great addition, and I have no objections to the approach. 🙂

Drenmi commented:
> should blocks be treated as valid for class macros?

Off the top of my head, I can only think of positive effects from doing this. (I.e. rooting out false negatives.)

backus commented:
@Drenmi updated the code, added docs, updated the spec style to be consistent with the file, and expanded the scope of `#macro?` to cover blocks too :smile:

backus commented:
Ah I found another case where a different cop wants to also respect access modifiers inside of `begin ... end` chunks. https://github.com/bbatsov/rubocop/blob/04af3f3353079eaad1918b603959408216e8c84e/spec/rubocop/cop/lint/useless_access_modifier_spec.rb#L481-L500

This PR is a nice little example of why it is nice to consolidate these predicates I guess. Removes random entropy between different cops

backus commented:
Alright it supports `kwbegin` nodes that would otherwise be inside a macro scope now. I had to refactor a few methods in one of the access modifier cops since my change set off one of the AbcSize metrics

backus commented:
I think this encapsulation of "what is an access modifier" is definitely a good thing. It is interesting though because a lot of conditionals had to go from

```
if cop_specific_access_modifier_predicate?(node)
  ...
end
```

to

```
if node.send_type? && node.access_modifier?
end
```

The fact that all of these type specific predicates worked with all nodes before was probably bad, but it does feel weird for the refactor PR to introduce a bit of clumping of these two predicates. Pretty sure it is fine but a small part of me wants to add

```
def access_modifier?
  false
end
```

to the base `Node` class.

Drenmi commented:
> The fact that all of these type specific predicates worked with all nodes before was probably bad, but it does feel weird for the refactor PR to introduce a bit of clumping of these two predicates.

Yes. I think about this quite a lot as well. We essentially have a typed system at the moment. I have implemented polymorphic interfaces for some related node types, but far from all of them.

In a lot of cases, we can be certain of the node type we get, e.g. in `on_*` callbacks, when using matchers. But in some cases it can also be a hodgepodge, e.g. when iterating over nodes in a method body.

(I have also identified some other common patterns, e.g. `node.parent && node.parent.foo_type?` where it could be nice to have a null object.)

When it comes to implementing no-op versions of all methods on `Node`, I am on the fence. I haven't thought about it enough. I appreciate the discussion being brought up. 🙂

backus commented:
k this build should finally be green @Drenmi :smile:

Drenmi commented:
@backus: Just came across a related thought. Should top level scope also be considered `#macro?`

backus commented:
Yeah probably!

backus commented:
@Drenmi `#macro?` now supports top level scopes too. I had to update some other tests to make them compatible with this change.

This should be ready for review :smile:

bbatsov commented:
👍
